File: python-algorithm/leetcode/problem_318.py
"""318. Maximum Product of Word Lengths
https://leetcode.com/problems/maximum-product-of-word-lengths/
"""
from typing import List


class Solution:
    def max_product(self, words: List[str]) -> int:
        def has_common(ws1, ws2) -> bool:
            if len(ws1) > len(ws2):
                return has_common(ws2, ws1)
            for ch in ws1:
                if ch in ws2:
                    return True
            return False

        ans = 0
        words_set = []
        for word in words:
            words_set.append(set(word))
        for i in range(len(words_set)):
            for j in range(i + 1, len(words_set)):
                if not has_common(words_set[i], words_set[j]):
                    ans = max(ans, len(words[i]) * len(words[j]))
        return ans

File: java-algo/src/main/java/com/leetcode/Problem318.java
package com.leetcode;

/**
 * 318. Maximum Product of Word Lengths
 * https://leetcode.com/problems/maximum-product-of-word-lengths/
 */
public class Problem318 {
    public int maxProduct(String[] words) {
        int ans = 0;
        int[][] store = new int[words.length][26];
        for (int i = 0; i < words.length; i++) {
            for (int j = 0; j < words[i].length(); j++) {
                store[i][words[i].charAt(j) - 'a'] += 1;
            }
        }
        for (int i = 0; i < words.length; i++) {
            for (int j = i + 1; j < words.length; j++) {
                if (helper(store[i], store[j])) {
                    ans = Math.max(ans, words[i].length() * words[j].length());
                }
            }
        }
        return ans;
    }

    private boolean helper(int[] a, int[] b) {
        for (int i = 0; i < 26; ++i) {
            if (a[i] > 0 && b[i] > 0) {
                return false;
            }
        }
        return true;
    }
}
